### PR TITLE
Enhance User Experience with Input Form and Email Notifications

### LINKED ISSUE #18: Disallow Users to Unsubscribe if Not Subscribed
Users should not be able to unsubscribe if they're not subscribed in the first place. There's a couple ways to fix this:

- Add a check to see if Datastore can find the user trying to unsubscribe and still redirect them to the home page
- Add a check to see if Datastore can find the user trying to unsubscribe and if not, bring up a text box/modal that says user can not be added because not in the system

image

### LINKED ISSUE #28: Add Custom Error Message for Emails
In the `sign-up.html` form, users are strictly limited to @google.com emails since this app is for internal use only but it doesn't specify which email to use currently.

image

### LINKED ISSUE #37: Include HTML Functionality in Emails
Emails are sent in direct text but can be styled to more with HTML content.

### PR DESCRIPTION
**Pull request by areetaw**
**Disallow Users to Unsubscribe if Not Subscribed**
Resolves #18. Previously, users were able to update their preferences of `subscribing` vs `unsubscribing` through a boolean. However, this brought up the issue of letting users `unsubscribing` without being in the database in the first place and keeping their information even after they have subscribed. Now, there will be a check in place to see if users are in our database and will delete them after they unsubscribe.

**Add Custom Error Message for Emails**
Resolves #28. When inputting emails for notifications, users had to follow a strict format of `+@google.com` and were notified through a general error message. Now, users will be able to see an additional error message specifying the exact format!
image

**Include HTML Functionality in Emails**
Resolves #37. Instead of plain text, HTML functionality is added to look more appealing to users. Additionally, when users sign up, they will be welcomed with an email letting them know they've subscribed.

image

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 1]
```diff
@@ -32,26 +38,39 @@
...
+      // Send a welcome email.
+      GmailConfiguration.sendEmail(email, Email.welcomeSubject, Email.welcomeContent);	
+    }
 
-    // Datastores updates the entity if it existed before based on email key
-    datastore.put(userEntity);
     response.sendRedirect("/index.html");
   }
-}
+}
```
Comment (mirrorkeydev) [inline src/main/java/com/google/sps/servlets/UserDataServlet.java:76]: Github shows newline missing (red circle w line) :0

File: src/main/java/com/google/sps/data/Email.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  public static final String welcomeContent = "<body style=\"margin: 3rem\">" +
+    "<h1>welcome to noms</h1>" +
+    "<p> thank you so much for subscribing to our mailing list of free food!";
+
+  public static final String dailyDigestContent = "<body style=\"margin: 3rem\">" +
+    "<h1>daily digest of free food @ university of california-irvine</h1>" +
+    "<div style=\"background: green; border-radius: 15px;" + 
+    "<p>pad thai @ aldrich park</p>";
+
+}
```
Comment (ubahl) [inline src/main/java/com/google/sps/data/Email.java:31]: Haha amazing!

File: src/main/java/com/google/sps/servlets/UserDataServlet.java (hunk 1 above)
Comment (areetaw) [reply src/main/java/com/google/sps/servlets/UserDataServlet.java:76]: Gotcha!

File: src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 3]
```diff
@@ -32,26 +38,39 @@
...
-    userEntity.setProperty("name", name);
-    userEntity.setProperty("university", university);
-    userEntity.setProperty("subscribe", subscribe);
+      // Remove user if in database.
+      Key userKey = KeyFactory.createKey("User", email);
+      try {
+        datastore.get(userKey);
+        datastore.delete(userKey);
+      } catch (EntityNotFoundException e) {
+        System.out.println("ERROR: User wants to unsubscribe but was never subscribed in the first place.");
```
Comment (mrboxtobox) [inline src/main/java/com/google/sps/servlets/UserDataServlet.java:58]: For now, this is okay, but ideally you want to use a logger for such errors.

Comment (areetaw) [reply src/main/java/com/google/sps/servlets/UserDataServlet.java:58]: Gotcha! Just added a logger!

**Pull request merged**